Add new workflow to automatically (and periodically) update `ex_doc` version

MarkoMin opened the pull request:
Tried everything locally (except actually creating an issue). If it works correctly, it will create one at midnight. First time having fun with GH actions so there may be some mistakes.

Closes #57

paulo-ferraz-oliveira commented on .github/workflows/ex_doc_version_check.yml:
```diff
@@ -0,0 +1,31 @@
+      - uses: actions/checkout@v2
+      - name: Get current version
+        run: |
+            CURR=`curl -s https://api.github.com/repos/elixir-lang/ex_doc/tags | \
+                  grep "\"name\":" | \
+                  sort --reverse | \
+                  head -n 1 | \
+                  awk '{print substr($2, 2, length($2)-3)}'`
```
This is probably safe, but if `ex_doc` ever releases a patch to an older version, there's no guarantee the latest tag is the "highest" in terms of versioning...

MarkoMin replied:
Well, yes, but as long as vMajor.minor.patch naming convention is used, it should be safe. We can utilize grep to only match versions which follow the convention

paulo-ferraz-oliveira replied:
(mind you this API call's response is also paginated, getting only the first 30 results, as-is)

MarkoMin replied:
Hmm, do we need more than the first page? But yeah, I completely missed pagination in this PR

paulo-ferraz-oliveira replied:
> more than the first page?

Theoretically, yes; in practice I don't know. I just wanted to point out a potential issue.

MarkoMin replied:
I can’t come up with an example where we would need them. I think we’re safe as long as the first page contains newest tags and if ex_doc doesn’t publish 2 releases in a day (in which case we’d miss one)

paulo-ferraz-oliveira replied:
The example would be `ex_doc` to publish 30+ support versions for older tags (when paginating GitHub doesn't care about semver or whatever is "highest", just "more recent"). In that case (unlikely, but still...) even if your comparison were using semver it would fail to recognise a new version. And since I can't be sure there won't be 30+ support versions... now that I think about it, is there no way to get "latest"? That'd solve the pagination issue, too.

MarkoMin replied:
Then we can increase per-page entries to considerably large number (e.g. 100) and sort tags by their name.

MarkoMin replied:
okay, per-page parameter can me from 1-100, but I think it should do the job even if is not perfect...  Worst it can do is a false-positive issue that can be closed

paulo-ferraz-oliveira replied:
I'm Ok with that.

paulo-ferraz-oliveira commented on .github/workflows/ex_doc_version_check.yml:
```diff
@@ -0,0 +1,31 @@
+                  head -n 1 | \
+                  awk '{print substr($2, 2, length($2)-3)}'`
+            TITLE="\`ex_doc\` ${CURR} released"
+            ISSUE_EXISTS=`curl -s https://api.github.com/repos/starbelly/rebar3_ex_doc/issues | grep -c "\"title\": \"${TITLE}\""`
```
This is also probably safe, if GitHub issues for `rebar3_ex_doc` don't "paginate" over a given limit, which is why I proposed the `EX_DOC_VSN` file.

MarkoMin replied:
Absolutely. We can solve this by fixing GH API version via header. I can submit modifications next week

paulo-ferraz-oliveira replied:
> fixing GH API version via header

Does that prevent pagination? What I meant what that if you have too many issues, it's possible that the API won't give you all of them on the response and them you'll assume there's no issue (as per what you're searching) when there actually is one.

MarkoMin replied:
I think yes. Currently docs dont mention any pagination, but we can try and fetch issues from some repo that have few k issues just to be sure. So current version of API should be supported and work consistently for a few years

paulo-ferraz-oliveira replied:
Yeah, doing this request on `erlang/otp`, as an example, brings up 30 results, while (atm) 275 are present...

paulo-ferraz-oliveira replied:
One idea might be to tag the issue and then retrieve it based on that? e.g. tag it with `_ ex_doc_version_check` and then retrieve it with `labels= ex_doc_version_check` (?)

paulo-ferraz-oliveira replied:
They'd have to abuse the label alongside the title, right? I'm Ok with your current solution, though.

paulo-ferraz-oliveira commented on .github/workflows/ex_doc_version_check.yml:
```diff
@@ -0,0 +1,31 @@
+
+            if [ ${ISSUE_EXISTS} -ne 0 ]
+            then
+              RESP=`curl -L \
+                -X POST \
+                https://api.github.com/repos/starbelly/rebar3_ex_doc/issues \
+                -d '{"title":"${TITLE}","body":"This is an automated report."}'`
```
I wonder if this won't run in all forks, too (?)

Is there a way to further restrict this workflow? Otherwise forks either have to disable it by hand or have their "Actions" tab with issues every now and then (I'm assuming they can't just post an issue as simply as this, but have to use some kind of token).

MarkoMin replied:
Yes, I believe it would. Prolly some env variable with current repo name exist, so we can wrap the whole task into IF clause.

paulo-ferraz-oliveira replied:
Or even better, not run the workflow in forks at all, otherwise it still pollutes the forks' tabs (though it might not be important 🤷).

paulo-ferraz-oliveira replied:
It seems this changed recently, and now actions are opt-in, so I'll go ahead and resolve this one 👍 

<img width="737" alt="image" src="https://github.com/starbelly/rebar3_ex_doc/assets/7303768/df496286-4d42-4f6f-a164-5ffcdffb3042">

paulo-ferraz-oliveira commented:
@MarkoMin, I created an Elixir script your pull request could plug in to (if it gets merged first), so that a pull request also gets created alongside the issue. 👍 

My pull request is mentioned above.

And here's an example of an automated pull request: https://github.com/paulo-ferraz-oliveira/rebar3_checkshell/blob/main/.github/workflows/rebar3_depup.sh (I do prefer to have my shell script separate so I can "shellcheck" and "shfmt" it 😄).

MarkoMin commented:
Oh, I didn't even think about that, that could work nicely! I'll utilize it if/when it gets merged.

MarkoMin commented:
One thought:  do we actually need both an issue and a PR? We should switch to PR only IMO.

paulo-ferraz-oliveira commented:
I thought about that, too. Maybe PR alone is enough.

MarkoMin commented:
Seems like it should be. But now we wait for #67 to get merged

paulo-ferraz-oliveira commented:
@MarkoMin, #67's merged. It ended up being a Mix task. I think you'd need to compile it and run it, unless @starbelly has a different suggestion.

paulo-ferraz-oliveira commented on .github/workflows/ex_doc_version_check.yml:
```diff
@@ -0,0 +1,31 @@
+on:
+  schedule:
+    - cron: '0 0 * * *'  # Run daily at midnight
+
```
```suggestion
  workflow_dispatch:
  
```

paulo-ferraz-oliveira replied:
This is to let maintainers run the workflow, from e.g. `main` by pressing a button and choosing a branch.

paulo-ferraz-oliveira commented on .github/workflows/ex_doc_version_check.yml:
```diff
@@ -0,0 +1,31 @@
+      - name: Get current version
+        run: |
+            CURR=`curl -s https://api.github.com/repos/elixir-lang/ex_doc/tags?per_page=100 | \
+                  grep "\"name\":" | \
+                  sort --reverse | \
+                  head -n 1 | \
+                  awk '{print substr($2, 2, length($2)-3)}'`
```
```suggestion
                  jq -r ".[0].name"
```

paulo-ferraz-oliveira replied:
`jq` is already installed in the Ubuntu runners...

paulo-ferraz-oliveira replied:
As-was, though, this was returning `v0.9.0`, most probably due to the `sort`. Maybe since the list returns the latest tag as the first tag we should just get `per_page=1` (?)

paulo-ferraz-oliveira commented on .github/workflows/ex_doc_version_check.yml:
```diff
@@ -0,0 +1,31 @@
+    name: Check ex_doc version
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
```suggestion
      - uses: actions/checkout@v4
```

paulo-ferraz-oliveira commented on .github/workflows/ex_doc_version_check.yml:
For consistency, we should probably name this workflow `up_ex_doc_version.yml`.

starbelly commented:
A thought on this, and this may or may not be pragmatic, but we have both erlang and elixir at our disposal. We could write a bit of code that uses httpc to get this information and handle the result with more exactness (i.e., use verl or Version, Version if it's a mix task). 

Other thought, since we're wanting to get the latest release that is available on hex, we can just hit the hex api to get this info. 

```
$ latest_stable=$(curl -s https://hex.pm/api/packages/ex_doc | jq '.latest_stable_version' | tr -d '"')
$ echo $latest_stable
0.30.6
```

paulo-ferraz-oliveira commented:
Good stuff. And `jq` has builtin support to remove the `"` too, so you can do `jq -r '.latest_stable_version'`. I'll suggest this above, instead of what I did before. I didn't quite get the Erlang idea, but feel free to expand. (e.g. you'd want the mix task to have no inputs and figure out the latest by itself?)

paulo-ferraz-oliveira commented on .github/workflows/ex_doc_version_check.yml:
```diff
@@ -0,0 +1,31 @@
+      - uses: actions/checkout@v2
+      - name: Get current version
+        run: |
+            CURR=`curl -s https://api.github.com/repos/elixir-lang/ex_doc/tags?per_page=100 | \
```
```suggestion
            CURR=`curl -s https://hex.pm/api/packages/ex_doc | jq -r '.latest_stable_version' \
```

paulo-ferraz-oliveira replied:
As per @starbelly's comment on the main conversation.

starbelly commented:
> Good stuff. And `jq` has builtin support to remove the `"` too, so you can do `jq -r '.latest_stable_version'`. I'll suggest this above, instead of what I did before. I didn't quite get the Erlang idea, but feel free to expand. (e.g. you'd want the mix task to have no inputs and figure out the latest by itself?)

Yeah. More like you can use erlang or elixir and httpc to do the gets and posts vs a bunch of shell scripterty, but that said. Let’s be pragmatic.

MarkoMin commented:
I've made following changes:

1. Moved task from root to `lib/mix/tasks`
2. Renamed workflow as per @paulo-ferraz-oliveira suggestion
3. Update bash script to make an automated PR as per @paulo-ferraz-oliveira example - take a look if there are any commands that are redundant or missing

paulo-ferraz-oliveira commented:
I approved the workflows to run. I'm forking your branch to test out the flow...

paulo-ferraz-oliveira commented:
@MarkoMin, you copied to "lib/mix/tasks", instead of moving...

paulo-ferraz-oliveira commented:
Example of a pull request working: https://github.com/paulo-ferraz-oliveira/rebar3_ex_doc_mm/pull/2/files.

paulo-ferraz-oliveira commented on .github/workflows/up_ex_doc_version.yml:
```diff
@@ -0,0 +1,49 @@
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - name: Get current version
```
```suggestion
      - uses: erlef/setup-beam@v1
        with:
          otp-version: 24
          elixir-version: 1.13
      - name: Get current version
```

paulo-ferraz-oliveira replied:
Without this `mix` isn't available (it's used below).

Also, FYI, I picked the same Erlang/Elixir versions as from the other workflow.

paulo-ferraz-oliveira commented on .github/workflows/up_ex_doc_version.yml:
```diff
@@ -0,0 +1,49 @@
+      - uses: actions/checkout@v4
+      - name: Get current version
+        run: |
+            CURR=`curl -s https://hex.pm/api/packages/ex_doc | jq -r '.latest_stable_version' \
+                  jq -r ".[0].name"
```
```suggestion
            CURR=$(curl -s https://hex.pm/api/packages/ex_doc | jq -r '.latest_stable_version')
```

paulo-ferraz-oliveira replied:
This command was incomplete (no closing "`"). Maybe it was a suggestion of mine, but here it's fixed.

paulo-ferraz-oliveira commented on .github/workflows/up_ex_doc_version.yml:
```diff
@@ -0,0 +1,49 @@
+        run: |
+            CURR=`curl -s https://hex.pm/api/packages/ex_doc | jq -r '.latest_stable_version' \
+                  jq -r ".[0].name"
+            TITLE="\`ex_doc\` ${CURR} released"
```
```suggestion
            TITLE="[automation] Update \`ex_doc\` to ${CURR}"
```

paulo-ferraz-oliveira commented on .github/workflows/up_ex_doc_version.yml:
```diff
@@ -0,0 +1,49 @@
+            CURR=`curl -s https://hex.pm/api/packages/ex_doc | jq -r '.latest_stable_version' \
+                  jq -r ".[0].name"
+            TITLE="\`ex_doc\` ${CURR} released"
+            PR_EXISTS=`curl -s https://api.github.com/repos/starbelly/rebar3_ex_doc/pulls?state=all%26per_page=100 | grep -c "\"title\": \"${TITLE}\""`
```
```suggestion
            PR=$(curl -s https://api.github.com/repos/${GITHUB_REPOSITORY}/pulls?state=all%26per_page=100 | jq ".[] | select(.title==\"${TITLE}\")")
```

paulo-ferraz-oliveira replied:
I ease running in forks (by means of `$GITHUB_REPOSITORY`) and also use `jq` instead of `grep`.

paulo-ferraz-oliveira commented on .github/workflows/up_ex_doc_version.yml:
```diff
@@ -0,0 +1,49 @@
+            TITLE="\`ex_doc\` ${CURR} released"
+            PR_EXISTS=`curl -s https://api.github.com/repos/starbelly/rebar3_ex_doc/pulls?state=all%26per_page=100 | grep -c "\"title\": \"${TITLE}\""`
+
+            if [ ${PR_EXISTS} -ne 0 ]
```
```suggestion
            echo "Searching for pull request with title ${TITLE}..."
            
            if [ -z "${PR}" ]
```

paulo-ferraz-oliveira replied:
Logging and minor change as per what was suggested above...

paulo-ferraz-oliveira commented on .github/workflows/up_ex_doc_version.yml:
```diff
@@ -0,0 +1,49 @@
+            then
+              git config user.name "GitHub Actions"
+              git config user.email "actions@user.noreply.github.com"
+
```
```suggestion
              echo "  ... not found in 100 most recent pull requests!"
```

paulo-ferraz-oliveira replied:
Logging...

paulo-ferraz-oliveira commented on .github/workflows/up_ex_doc_version.yml:
```diff
@@ -0,0 +1,49 @@
+              git config user.email "actions@user.noreply.github.com"
+
+              BRANCH="up_ex_doc_to_${CURR}"
+
```
```suggestion
              git fetch origin
              
```

paulo-ferraz-oliveira replied:
If we don't do it here branch comparison is wrong.

paulo-ferraz-oliveira commented on .github/workflows/up_ex_doc_version.yml:
```diff
@@ -0,0 +1,49 @@
+              git config user.name "GitHub Actions"
+              git config user.email "actions@user.noreply.github.com"
+
+              BRANCH="up_ex_doc_to_${CURR}"
```
```suggestion
              git fetch origin
              
              BRANCH="up_ex_doc_to_${CURR}"
```

paulo-ferraz-oliveira replied:
If we don't do it (`fetch`), here, branch comparison is wrong.

paulo-ferraz-oliveira commented on .github/workflows/up_ex_doc_version.yml:
```diff
@@ -0,0 +1,49 @@
+              BRANCH="up_ex_doc_to_${CURR}"
+
+              if git branch -a | grep "$BRANCH" > /dev/null; then
+                  # already exists
```
```suggestion
                  echo "Branch $BRANCH found! Bailing out..."
```

paulo-ferraz-oliveira replied:
Logging...

paulo-ferraz-oliveira commented on .github/workflows/up_ex_doc_version.yml:
```diff
@@ -0,0 +1,49 @@
+                  exit
+              fi
+
+              git fetch origin
```
```suggestion
```

paulo-ferraz-oliveira replied:
Moved above.

paulo-ferraz-oliveira commented on .github/workflows/up_ex_doc_version.yml:
```diff
@@ -0,0 +1,49 @@
+
+              git fetch origin
+              git checkout -b "$BRANCH"
+              mix compile
```
```suggestion
              mix deps.get
              mix compile
```

paulo-ferraz-oliveira replied:
Without `mix deps.get` this fails...

paulo-ferraz-oliveira commented on .github/workflows/up_ex_doc_version.yml:
```diff
@@ -0,0 +1,49 @@
+              mix up_ex_doc_version "${CURR}"
+
+              if ! git diff --exit-code 1> /dev/null ; then
+                  # there's stuff to push
```
```suggestion
                  echo "There's stuff to push. Creating a pull request..."
```

paulo-ferraz-oliveira replied:
Logging...

paulo-ferraz-oliveira commented on .github/workflows/up_ex_doc_version.yml:
```diff
@@ -0,0 +1,49 @@
+              if ! git diff --exit-code 1> /dev/null ; then
+                  # there's stuff to push
+                  git add .
+                  git commit -m "Update ex_doc version"
```
```suggestion
                  git commit -m "${TITLE}"
```

paulo-ferraz-oliveira replied:
Consistency...

paulo-ferraz-oliveira commented on .github/workflows/up_ex_doc_version.yml:
```diff
@@ -0,0 +1,49 @@
+                  git push origin "$BRANCH"
+
+                  gh pr create --fill \
+                      --title "Update ex_dpc version (automation)" \
```
```suggestion
                      --title "${TITLE}" \
```

paulo-ferraz-oliveira replied:
Consistency...

paulo-ferraz-oliveira commented on .github/workflows/up_ex_doc_version.yml:
```diff
@@ -0,0 +1,49 @@
+
+                  gh pr create --fill \
+                      --title "Update ex_dpc version (automation)" \
+                      --body "This is an automated action to update ex_doc version"
```
```suggestion
                      --body "This is an automated action to update \`ex_doc\` to version ${CURR}"
```

paulo-ferraz-oliveira commented on .github/workflows/up_ex_doc_version.yml:
```diff
@@ -0,0 +1,49 @@
+                  gh pr create --fill \
+                      --title "Update ex_dpc version (automation)" \
+                      --body "This is an automated action to update ex_doc version"
+              fi
```
```suggestion
              else
                  echo "Nothing to push. Is \`ex_doc\` already at version ${CURR}?"
              fi
            else
              echo "  ... found! Bailing out..."
```

paulo-ferraz-oliveira commented on .github/workflows/up_ex_doc_version.yml:
```diff
@@ -0,0 +1,49 @@
+                      --title "Update ex_dpc version (automation)" \
+                      --body "This is an automated action to update ex_doc version"
+              fi
+            fi
```
```suggestion
            fi
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

paulo-ferraz-oliveira replied:
Without this you can't push commits or create pull requests. More info. here: https://docs.github.com/en/actions/security-guides/automatic-token-authentication

paulo-ferraz-oliveira commented on .github/workflows/up_ex_doc_version.yml:
```diff
@@ -0,0 +1,49 @@
+    steps:
+      - uses: actions/checkout@v4
+      - name: Get current version
+        run: |
```
```suggestion
```

paulo-ferraz-oliveira commented on lib/mix/tasks/up_ex_doc_version.ex:
```diff
@@ -0,0 +1,30 @@
+      |> run(new_vsn)
+
+    File.write!(@mix_exs, "#{content_up}\n")
+    Mix.Task.run("deps.unlock", ["ex_doc"])
+    Mix.Task.run("deps.get")
```
```suggestion
    env = [{"PATH", System.get_env("PATH")}]
    System.cmd("mix", ["deps.unlock", "ex_doc"], env: env)
    System.cmd("mix", ["deps.get"], env: env)
```

paulo-ferraz-oliveira replied:
I had an off-by-one issue here. Mix's state is not updated when we write the file, so `Mix.Task.run` would only work on the second update.

paulo-ferraz-oliveira commented on lib/mix/tasks/up_ex_doc_version.ex:
- [ ] you need to remove `up_ex_doc_version.ex` at the root folder.

paulo-ferraz-oliveira commented:
@MarkoMin, sorry I did so many small changes/suggestions, but it was the best way I found to explain my reasoning. Accepting them all should put the pull request in a mergeable state.

starbelly commented:
> @starbelly, if this isn't done already we're gonna need to enable
> <img alt="image" width="663" src="https://user-images.githubusercontent.com/7303768/272067604-109f4a43-5288-49d9-8040-e66943e40a10.png">
> 
> and also
> <img alt="image" width="497" src="https://user-images.githubusercontent.com/7303768/272071239-cba9a19b-dcb6-4f7b-8df0-331208ea33dc.png">

I think with how infrequent ex_doc releases are (and should be) we should just allow the bot to open PR. We definitely want to see what the bot has done before merging, thoughts?

paulo-ferraz-oliveira commented:
That's my goal. I was under the impression you needed:

- the latter to allow to write to branch
- the former to open the pull request

But maybe the former is not required (?)

paulo-ferraz-oliveira commented:
We can also accept+merge (once all changes are done), then test the settings that are required (our goal is only to create the pull request, as automating stuff like this can lead to issues - let's say that `ex_doc` breaks and we don't have enough tests, for example).

MarkoMin commented:
I approved all suggestions. Sorry for errors with `mix`, first time using it. 

Now it should be done. It's quite messy with all those commits, but we can squash them I guess. For some reason, `mix.exs` got into commits, but changed version is the file that is currently on main so no conflicts should exists.

starbelly commented on lib/mix/tasks/up_ex_doc_version.ex:
```diff
@@ -0,0 +1,31 @@
+
+    File.write!(@mix_exs, "#{content_up}\n")
+    env = [{"PATH", System.get_env("PATH")}]
+    System.cmd("mix", ["deps.unlock", "ex_doc"], env: env)
```
We can do this and it should be fine, but probably more pragmatic is to run this, then execute `mix deps.get`  back in the shell (i.e., in whatever workflow that will use this). 

Thoughts?

paulo-ferraz-oliveira replied:
I don't mind doing it in another pull request @starbelly, even before we release. This fix is already scope creep caused by me, since it wasn't in the change proposed by @MarkoMin initially.

paulo-ferraz-oliveira replied:
@MarkoMin, if you prefer to do it, you can remove these calls to mix from here and set them in the CI bit... But lemme know, otherwise I can also do in a subsequent pr (since this one's working as-is)

MarkoMin replied:
I can do it. Do I need to remove File:write also or just lines 14 and 15?

MarkoMin replied:
done

paulo-ferraz-oliveira commented:
@MarkoMin, could you rebase this onto the main branch, and move the script?

I see this in the Files' tab

<img width="326" alt="image" src="https://github.com/starbelly/rebar3_ex_doc/assets/7303768/5db61245-3d50-4c7d-a162-9f1bc8e6e6f1">

and it's confusing. Thanks.

MarkoMin commented:
> @MarkoMin, could you rebase this onto the main branch, and move the script?

done

starbelly reviewed (approved):
LGTM! If it fails, we can fix it 😄
